### ISSUE TITLE
Implement pausing

### ISSUE THREAD

**Issue by benfrankel** [TheBevyFlock/bevy_new_2d#206]
How this can be implemented:

- Define a state `struct Pause(bool)` somewhere (`core/pause.rs` if we had `core/`, otherwise probably just `src/pause.rs` or I guess `src/lib.rs`).
- Every system that ought to not run while the game is paused should add `.run_if(not(in_state(Paused(true))))` on a case-by-case basis (this could be made more ergonomic with a custom run condition or system set provided in `pause.rs`).
- For a pause menu, we'd need a separate state, probably a sub-state of `Screen::Playing`, and toggle `Pause` when you enter / exit that state. In my jam game I had an `enum PlayingMenu { Pause, LevelUp, ... }`.

**Comment by janhohenheim** [comment]
<img width="737" alt="image" src="https://github.com/user-attachments/assets/79f26058-5df0-45c5-9434-c6be10041857">

We should probably replace the default behavior of `Esc` by pause while we are on it.

**Comment by TheKnarf** [comment]
Recently used this template for a game jam, and this was the one thing I felt was missing / was wrong from the template.

I think the screen state should probably be a stack! If you want to pause the game you should push a pause state on top of the stack. If you then want to navigate to a settings menu you can push that as well on top of the stack. Then by leaving the settings menu you just pop it from the stack, and then to leave the pause state you pop it from the stack.

That way you can re-use a settings menu from both the start screen and from a main menu (without having any special logic in the settings menu).

**Comment by TheKnarf** [comment]
Would it work to create your own type using a resource that tracks the stack. And two states that track the top and bottom of the stack with a system syncs the two states with the state of the stack?

**Comment by benfrankel** [comment]
Yeah, in my own project I would agree. The (not-yet-fully-codified) design principles of this template include 1. not using any 3rd-party dependencies besides `bevy`, and 2. not introducing complexity that a relative beginner may struggle to understand, unless it's _strictly_ necessary.

**Comment by TheKnarf** [comment]
>and 2. not introducing complexity that a relative beginner may struggle to understand, unless it's strictly necessary. So our options are limited by those requirements.
>
>Improving bevy upstream is the best way to improve this template while keeping it simple.

Feels like being able to support both a pause state and a nested settings menu is strictly necessary, but I do get your point. I think the template should take on the necessary complexity to support to support at least as much. I'd be happy to make PR, but I won't bother spending time on it if your unlikely accept it.

**Comment by janhohenheim** [comment]
@TheKnarf I would not submit a PR for this yet in your position. I'll keep this in the back of my mind though and talk about it at some point with the Bevy maintainers. 

I also want to see a nested settings menu supported, and if I get my way, we *may* really need a stack.

**Comment by will-hart** [comment]
I feel like we could get most of the way with just sub-states (although it may be a little ugly). For instance here is a quick attempt at a pause screen https://github.com/TheBevyFlock/bevy_quickstart/compare/main...will-hart:bevy_quickstart:pause-state

**Comment by janhohenheim** [comment]
@will-hart can I be inside multiple substates of the playing screen at once?

**Comment by benfrankel** [comment]
If they're different types then yes. Keep in mind a pause menu is actually two features though, pausing the game and opening a menu, which should use orthogonal state types.
